### PR TITLE
bulk permission update

### PR DESCRIPTION
**Pull request by jcardus**

### CODE REVIEW

File: src/main/java/org/traccar/api/resource/PermissionsResource.java [hunk 1]
```diff
@@ -17,7 +17,13 @@
 package org.traccar.api.resource;
 
 import java.sql.SQLException;
+import java.util.HashSet;
 import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
```
Comment (tananaev) [inline src/main/java/org/traccar/api/resource/PermissionsResource.java:26]: Let's not use stream API. We are not using it yet and it's probably not really needed here.

File: src/main/java/org/traccar/api/resource/PermissionsResource.java [hunk 2]
```diff
@@ -55,30 +61,61 @@ private void checkPermission(Permission permission, boolean link) {
...
     }
 
+    private void refreshPermissions(List<LinkedHashMap<String, Long>> entities) {
+        Set<String> ownerProperty = new HashSet<>();
+        entities.stream()
+            .map(e -> {
+                try {
+                    return new Permission(e);
+                } catch (ClassNotFoundException classNotFoundException) {
+                    classNotFoundException.printStackTrace();
```
Comment (tananaev) [inline src/main/java/org/traccar/api/resource/PermissionsResource.java:71]: Please use proper logging if needed.

File: src/main/java/org/traccar/api/resource/PermissionsResource.java [hunk 3]
```diff
@@ -55,30 +61,61 @@ private void checkPermission(Permission permission, boolean link) {
...
+                    return null;
+                }
+            })
+            .filter(Objects::nonNull)
+            .filter(p -> ownerProperty.add(p.getOwnerClass().getName() + p.getPropertyClass().getName()))
```
Comment (tananaev) [inline src/main/java/org/traccar/api/resource/PermissionsResource.java:76]: There has to be a better way of doing this based on the original `LinkedHashMap`.

File: src/main/java/org/traccar/api/resource/PermissionsResource.java [hunk 4]
```diff
@@ -55,30 +61,61 @@ private void checkPermission(Permission permission, boolean link) {
...
+            checkPermission(permission, false);
+            Context.getDataManager().linkObject(permission.getOwnerClass(), permission.getOwnerId(),
+                    permission.getPropertyClass(), permission.getPropertyId(), false);
+            LogAction.unlink(getUserId(), permission.getOwnerClass(), permission.getOwnerId(),
+                    permission.getPropertyClass(), permission.getPropertyId());
+        }
+        refreshPermissions(entities);
         return Response.noContent().build();
     }
-
```
Comment (tananaev) [inline src/main/java/org/traccar/api/resource/PermissionsResource.java:83]: revert

**Comment by jcardus** [comment]
would be easier to implement if we reload everything after the bulk update request

**Comment by tananaev** [comment]
Another option is to assume/enforce that all permissions are of the same type.

**Comment by jcardus** [comment]
you mean all with the same owner / property combination?

File: src/main/java/org/traccar/api/resource/PermissionsResource.java (hunk 4 above)
Comment (jcardus) [reply src/main/java/org/traccar/api/resource/PermissionsResource.java:83]: revert what?

**Comment by jcardus** [comment]
Can you check?

File: src/main/java/org/traccar/api/resource/PermissionsResource.java [hunk 5]
```diff
@@ -57,27 +59,48 @@ private void checkPermission(Permission permission, boolean link) {
...
-        Context.getPermissionsManager().refreshPermissions(permission);
+        for (LinkedHashMap<String, Long> entity: entities) {
+            Permission permission = new Permission(entity);
+            checkPermission(permission, true);
+            Context.getDataManager().linkObject(permission.getOwnerClass(), permission.getOwnerId(),
+                    permission.getPropertyClass(), permission.getPropertyId(), true);
+            LogAction.link(getUserId(), permission.getOwnerClass(), permission.getOwnerId(),
+                    permission.getPropertyClass(), permission.getPropertyId());
+        }
+        // we assume all permissions are of same type so we use the first one for refreshing
```
Comment (tananaev) [inline src/main/java/org/traccar/api/resource/PermissionsResource.java:77]: I think we should probably throw an exception if permissions are not the same type.

File: src/main/java/org/traccar/api/resource/PermissionsResource.java [hunk 6]
```diff
@@ -74,13 +76,25 @@ public Response add(List<LinkedHashMap<String, Long>> entities) throws SQLExcept
...
+        if (!entities.isEmpty()) {
+            Permission first = new Permission(entities.get(0));
+            for (LinkedHashMap<String, Long> entity: entities) {
+                Permission permission = new Permission(entity);
+                if (!first.getOwnerClass().equals(permission.getOwnerClass())
+                    || !first.getPropertyClass().equals(permission.getPropertyClass())) {
+                    throw new WebApplicationException(Response.status(Response.Status.BAD_REQUEST).build());
+                }
+            }
+        }
```
Comment (tananaev) [inline src/main/java/org/traccar/api/resource/PermissionsResource.java:95]: Can we do something more elegant like this:

```suggestion
        Set<String> keys = null;
        for (LinkedHashMap<String, Long> entity: entities) {
            if (keys != null & !entity.keySet().equals(keys)) {
                throw new WebApplicationException(Response.status(Response.Status.BAD_REQUEST).build());
            }
            keys = entity.keySet();
        }
```

Comment (jcardus) [reply src/main/java/org/traccar/api/resource/PermissionsResource.java]: ok

**Comment by jcardus** [comment]
can you check?

**Review by tananaev — COMMENTED** [review]
One last thing - can we order methods in a consistent manner. One way:

1. checkPermissionTypes
2. bulk add
3. add
4. bulk remove
5. remove

**Comment by tananaev** [comment]
By the way, let me know when it's ready for review. Just pushing updated code doesn't tell me that it's ready.

**Comment by jcardus** [comment]
please check

**Comment by tananaev** [comment]
Merged, thank you.

**Pull request merged**